### ISSUE TITLE
Refactoring needed in `database/extensions.dart`

### ISSUE THREAD

**Issue by lukehutch** [serverpod/serverpod#1677]
The code in `packages/serverpod/lib/src/database/extensions.dart` needs refactoring. Bath the big if-statement and the switch statement below it, selecting based on type, should be replaced with inheritance (it's a general observation that switching based on type is an anti-pattern).

```dart
/// SQL code generation methods for [FilterConstraint].
extension FilterConstraintGenerator on FilterConstraint {
  /// Generates a SQL WHERE clause from the filter constraint.
  String toQuery(ColumnDefinition columnDefinition) {
    assert(columnDefinition.name == column);

    var columnType = columnDefinition.columnType;
    String formattedValue;
    if (columnType == ColumnType.integer ||
        columnType == ColumnType.bigint ||
        columnType == ColumnType.doublePrecision ||
        columnType == ColumnType.boolean) {
      formattedValue = value;
    } else if (columnType == ColumnType.text) {
      formattedValue = DatabasePoolManager.encoder.convert(value);
    } else if (columnType == ColumnType.timestampWithoutTimeZone) {
      if (type == FilterConstraintType.inThePast) {
        formattedValue = _microsecondsToInterval(int.parse(value));
      } else {
        var dateTime = DateTime.tryParse(value);
        if (dateTime != null) {
          formattedValue = DatabasePoolManager.encoder.convert(dateTime);
        } else {
          formattedValue = 'NULL';
        }
      }
    } else {
      throw Exception(
        'Unsupported column type ${columnDefinition.columnType} for column '
        '"${columnDefinition.name}".',
      );
    }

    switch (type) {
      case FilterConstraintType.equals:
        return '"$column" = $formattedValue';
      case FilterConstraintType.notEquals:
        return '"$column" != $formattedValue';
      case FilterConstraintType.greaterThan:
        return '"$column" > $formattedValue';
      case FilterConstraintType.greaterThanOrEquals:
        return '"$column" >= $formattedValue';
      case FilterConstraintType.lessThan:
        return '"$column" < $formattedValue';
      case FilterConstraintType.lessThanOrEquals:
        return '"$column" <= $formattedValue';
      case FilterConstraintType.like:
        return '"$column" LIKE $formattedValue';
      case FilterConstraintType.notLike:
        return '"$column" NOT LIKE $formattedValue';
      case FilterConstraintType.iLike:
        return '"$column" ILIKE $formattedValue';
      case FilterConstraintType.notILike:
        return '"$column" NOT ILIKE $formattedValue';
      case FilterConstraintType.between:
        return '"$column" BETWEEN $formattedValue AND $value2';
      case FilterConstraintType.inThePast:
        return '"$column" > (NOW() - $formattedValue)';
      case FilterConstraintType.isNull:
        return '"$column" IS NULL';
      case FilterConstraintType.isNotNull:
        return '"$column" IS NOT NULL';
    }
  }
}
```

**Comment by Isakdl** [comment]
Both the if statements and the switch are made over enums. Why do you think this is an anti-pattern, could you expand on that?

**Comment by lukehutch** [comment]
There's some explanation of the thinking here, going back to Martin Fowler's thoughts on it:

https://wiki.c2.com/?SwitchStatementsSmell

However, the main reason I pointed this out is that you already have a `Column` class with many subclasses, so the `ColumnType` enum seems to duplicate that class taxonomy, and many of the `FilterConstraintType` types are specific to only one `Column` subclass (and others could be specific to a shared superclass that is a subclass of `Column`, for example all the scalar or ordinal types, including `DateTime`, could share a common ancestor class that would define the filtering behavior of the inequality operators).

Basically, in both cases, you're duplicating what you have already made object oriented in the `Column` hierarchy.

Case statements over enum values are not always indicative of code that should be rewritten into object oriented form, but as the link above explains, they do go against OO design when you need to do something different for all values (e.g. in a switch statement, or in this case in a class hierarchy) in more than one place in the program.

**Comment by Isakdl** [comment]
Alright got it, I'm generally a fan of Martin Fowler he usually has good takes. I think it depends on the use-case and the problem, I also think there are many other problems with OO and inheritance in general. But I digress it is irrelevant to this discussion.

Truth be told I have not looked at this piece of code before. I believe what is going on here is that we have some technical debt and duplicated code that has been left here after we refactored the query code. As far as I can tell this code is only used for the "insights" app and not when executing normal queries. When executing normal queries we build them with the internal query builder and we use the classes in the `database/columns.dart` file to construct the where queries.

The code in there very much follows your suggestion here.

Example:
```dart
class _BetweenExpression<T> extends _MinMaxColumnExpression<T> {
  _BetweenExpression(super.column, super.min, super.max);

  @override
  String get operator => 'BETWEEN';
}

class _NotBetweenExpression<T> extends _MinMaxColumnExpression<T> {
  _NotBetweenExpression(super.column, super.min, super.max);

  @override
  String get operator => 'NOT BETWEEN';
}
```

**Comment by Isakdl** [comment]
Perhaps I should add that you can find all our unit tests on the query builders in `packages/serverpod/test/database/**` and then our integration tests for database queries in `tests/serverpod_test_server/test_integration/column_operations` and `tests/serverpod_test_server/test_integration/database_operations`. These are pretty good at showing how the system work.

**Comment by lukehutch** [comment]
Ah, OK.

Thanks for the pointers -- I spent a lot of time getting to know the code back when I was trying to get columns to serialize to strings, but now a lot of things changed, so at this point I can't invest a lot of time into the internals anymore, I have to push forward with Serverpod as it is today.

You can close this issue if this code is not important. Thanks.
